### PR TITLE
ESLint: Enable `eslint-plugin-react-compiler`

### PR DESCRIPTION
**Pull request by tyxla**
## What?
This PR introduces the `eslint-plugin-react-compiler` ESLint plugin to the repo.

Even if the plugin is still beta, it will help us to have it in the repo in order to maintain a friendlier environment for React 19 and the React Compiler.

We're experimenting with adopting React Compiler through the Babel plugin in #66361.

## Why?

[It's recommended](https://react.dev/learn/react-compiler) that this ESLint plugin is used to help improve the quality of your codebase.

## How?
The `eslint-plugin-react-compiler` plugin can be used without the React Compiler itself. We're installing it and enabling it. Also ignoring a few files that include top-level `await` which seems to break the linter right now.

## Testing Instructions
Verify there are no ESLint errors and the static analysis check passes. 

### Testing Instructions for Keyboard
None

## Screenshots or screencast 
None

### CODE REVIEW

**Comment by tyxla** [comment]
It will trigger an ESLint error everywhere where we ignore a `react/*` ESLint rule:

```
React Compiler has skipped optimizing this component because one or more React ESLint rules were disabled. React Compiler only works when your components follow all the rules of React, disabling them may result in unexpected or incorrect behavior
```

Seems like it will try hard to let us follow all the rules, which can create quite some additional work 😐

**Comment by Mamaduka** [comment]
I kinda like the new stricter rules, but I agree it would take some time to bring the codebase into shape.

P.S. I'll push fixes for some low-hanging errors separately.

**Comment by tyxla** [comment]
> I kinda like the new stricter rules, but I agree it would take some time to bring the codebase into shape.
> 
> P.S. I'll push fixes for some low-hanging errors separately.

I'd love to help with those too. Please ping me for reviews!

**Comment by Mamaduka** [comment]
@tyxla, should we start a list of false positive errors that linter generates and then report those upstream?

## Example

* Updating `ref` value on event callbacks. Setting the `ref` value during the render isn't recommended, but it should be okay in event handlers. File - `packages/block-editor/src/components/inserter-list-item/index.js`.

**Comment by tyxla** [comment]
> @tyxla, should we start a list of false positive errors that linter generates and then report those upstream?
> 
> ## Example
> * Updating `ref` value on event callbacks. Setting the `ref` value during the render isn't recommended, but it should be okay in event handlers. File - `packages/block-editor/src/components/inserter-list-item/index.js`.

Yep, definitely. The top-level await could also be one to report (currently not occurring because I've ignored the files where it happens).

File: .eslintrc.js
Comment (Mamaduka) [inline .eslintrc.js]: We can probably ignore `.story.js` files and anything in `test` sub-directories.

Comment (gziolo) [reply .eslintrc.js]: This would help:

https://github.com/WordPress/gutenberg/blob/9a2d9a4f260ba2b684ce68f709ebdeeec3e61085/.eslintrc.js#L27-L32

Comment (tyxla) [reply .eslintrc.js]: Done in https://github.com/WordPress/gutenberg/pull/61788/commits/5921e1e402347f4f7e1f18c7a807c853fef0793b.

File: .eslintrc.js
Comment (gziolo) [inline .eslintrc.js]: Maybe we should have another group for Interactivity API, which uses Preact behind the scenes. It should be:

- `packages/interactivity-router/src/*`
- `packages/interactivity/src/*`

Comment (tyxla) [reply .eslintrc.js]: Makes sense. I guess we can use the `packages/interactivity*/src/**` existing group?

Comment (tyxla) [reply .eslintrc.js]: Done in 54f002a95c8a06b9449e44f30f79f8855b731d3e

**Comment by Mamaduka** [comment]
I noticed that new lint rules do not like the hook pattern used by the command API:

Example:

```js
useCommandLoader( {
  name: 'core/edit-site/reset-global-styles',
  hook: useGlobalStylesResetCommands,
} );
```

**Comment by tyxla** [comment]
> I noticed that new lint rules do not like the hook pattern used by the command API:

The solution to that will most likely be just renaming them to not follow the hook naming convention. After all, they are technically not hooks since they're violating the rules of hooks and React can't recognize them as hooks through static code analysis:

https://github.com/WordPress/gutenberg/blob/c56672504bada434aee1ec331366cd2a10ab2b38/packages/commands/src/components/command-menu.js#L38

**Comment by Mamaduka** [comment]
It looks like, after the latest update, the plugin complains about setting the `ref` value inside effects.

**Comment by tyxla** [comment]
> It looks like, after the latest update, the plugin complains about setting the `ref` value inside effects.

Yes, looks like it's a known bug: https://github.com/facebook/react/issues/29196

**Comment by Mamaduka** [comment]
I'm talking about different case that's is supported in React and ref is coming from local `useRef` hook. Example: https://github.com/WordPress/gutenberg/actions/runs/9305858965/job/25613775078?pr=61788#step:5:53.

I'm not able to reproduce it on React Compiler Playground though.

## Screenshot
CleanShot 2024-05-31 at 16 49 39

**Comment by tyxla** [comment]
Perhaps I'm misunderstanding, but isn't that precisely the bug reported in the issue I linked above?

**Comment by Mamaduka** [comment]
No, the linked bug report is about `ref` coming from another hook, instead of `useRef`.

```js
function useExampleRef() {
	const ref = useRef();
	return ref;
}

function Example1() {
	const ref = useExampleRef();

	useEffect( () => {
		ref.current = 'value';
	} );
}

// above /react/issues/29196 vs what linter flags in this PR.

function Example2() {
	const ref = useRef();

	useEffect( () => {
		ref.current = 'value';
	} );
}
```

**Comment by tyxla** [comment]
This happens because the React Compiler ESLint plugin doesn't recognize that this is React's `useRef` - we import it from `@wordpress/element`. 

If you wish to try it out, try changing:

```
import { useRef } from '@wordpress/element';
```

to:

```
import { useRef } from 'react';
```

and you'll see the error will be gone.

That being said, I still think it's more or less the same issue because for React Compiler it's all the same - it's a different hook from its perspective, as it can't understand that this is the same React `useRef`. 

Let me know if that makes sense.

**Comment by Mamaduka** [comment]
Oh, I didn’t though about that 😅

**Comment by tyxla** [comment]
This is a good reason to reconsider if we want to continue using `@wordpress/element` across the codebase, cc @youknowriad @jsnajdr

**Comment by tyxla** [comment]
Thanks for keeping it up to date, @Mamaduka. Bummer to see the `ref.current` change still caught as an unexpected mutation 👎

**Comment by Mamaduka** [comment]
The next Ariakit update should resolve any failures coming from it - https://ariakit.org/changelog#049.

**Comment by tyxla** [comment]
Working on the Ariakit update in https://github.com/WordPress/gutenberg/pull/64637. 

Seems like we found a bug and we'll wait for the next minor version that contains [the fix](https://github.com/ariakit/ariakit/pull/4056).

File: packages/eslint-plugin/configs/react.js
Comment (swissspidy) [inline packages/eslint-plugin/configs/react.js]:
```suggestion
		'react-compiler',
```

**Comment by swissspidy** [comment]
> This is a good reason to reconsider if we want to continue using `@wordpress/element` across the codebase

Or perhaps make an upstream contribution so that this is recognized by the lint rule.

File: packages/eslint-plugin/configs/react.js
Comment (tyxla) [reply packages/eslint-plugin/configs/react.js]: Thanks! Both should work, but the shorter one wins.

**Comment by tyxla** [comment]
> The next Ariakit update should resolve any failures coming from it - https://ariakit.org/changelog#049.

Indeed! Working on using the new `useStoreState` API that will resolve those errors in https://github.com/WordPress/gutenberg/pull/64648.

**Comment by tyxla** [comment]
With the Ariakit errors being fixed in #64648, it seems like the prevalent error now is this one:

```
Mutating a value returned from a function whose return value should not be mutated  react-compiler/react-compiler
```

Being a known issue (https://github.com/facebook/react/issues/29196), there are some interesting resolutions proposed, like [this one](https://github.com/reactwg/react-compiler/discussions/14) that proposes a `useSharedValue` hook. There have been some attempts to fix this (see https://github.com/facebook/react/pull/29916) but thus far, it seems it's still an issue.

**Comment by tyxla** [comment]
Made some progress today: 

1. Updated to the latest experimental version that includes the option to exclude ref mutation (https://github.com/facebook/react/pull/29916). Currently at **264** errors.
2. Updating to Ariakit's `useStoreState()` resolved 16 errors. Down to **252** errors.
3. Disabling `react-compiler` ESLint rule for React native files. Down to **140** errors.
4. Enabling `enableTreatRefLikeIdentifiersAsRefs` to allow React Compiler recognize refs as such through following specific name standards.
5. Renaming refs to the new convention. Down to **87** errors.

After a manual audit, I'm planning to land the ref renaming separately, as it's harmless. 

Next steps: there are some false positives that are worth auditing, plus a bunch of additional violations and `eslint-disable`s that need to be addressed one by one.

**Comment by Mamaduka** [comment]
Fantastic work, @tyxla!

> After a manual audit, I'm planning to land the ref renaming separately, as it's harmless.

Makes sense!

**Comment by Mamaduka** [comment]
This rule would be interesting to run against our codebase, probably in a separate PR - https://github.com/facebook/react/pull/30685.

**Comment by tyxla** [comment]
Here's a PR to land a rename of most of the refs: https://github.com/WordPress/gutenberg/pull/64718

**Comment by tyxla** [comment]
https://github.com/WordPress/gutenberg/pull/64718 landed, so I've removed its commit and rebased the PR once more.

**Comment by tyxla** [comment]
Fixing some unexpected `useInstanceId` hook references in #65733.

**Comment by tyxla** [comment]
Enabling React Compiler for `InitPatternModal` by getting rid of the unnecessary underlying `useEffect` in #65734.

**Comment by tyxla** [comment]
And another fix suggested in #65737, related to removing an `eslint-ignore` directive.

**Comment by tyxla** [comment]
Also, updated to the latest `0.0.0-experimental-f444e11-20240926` from last week.

**Comment by tyxla** [comment]
A bunch of these errors appear now in the latest version:

```
Ref values (the `current` property) may not be accessed during render (https://react.dev/reference/react/useRef)
```

Looks like this one will be suppressed soon: https://github.com/facebook/react/issues/30745#issuecomment-2329039708

**Comment by tyxla** [comment]
Fixing prop mutations in #65739 and #65740.

**Comment by tyxla** [comment]
Spent some brief time on it today, here's an update:

1. Started with **143** errors (more than they were before, see https://github.com/WordPress/gutenberg/pull/61788#issuecomment-2382962380).
2. Updated to the latest version from October 3rd (https://github.com/WordPress/gutenberg/pull/61788/commits/9cc6dab1d79fd692c17947a8bc6e6fe4aa699cdb).
3. Disabled `validateRefAccessDuringRender` in https://github.com/WordPress/gutenberg/pull/61788/commits/13bd705ed121a5f223eccbcc52846612e52b27b2, see https://github.com/facebook/react/pull/30843 for context.

This led to reducing errors back to **74**.

**Comment by tyxla** [comment]
Currently, part of the errors are because React Compiler doesn't recognize `@wordpress/element` as `react`. To fix this, we'll either need to start importing directly from `react`, or be able to specify `@wordpress/element` as a known React package. I'm going to play with the second idea in https://github.com/facebook/react/pull/31138.

**Comment by Mamaduka** [comment]
The React Compiler team just released beta versions and is looking for feedback from early adopters. It also supports React 17+, so we don't have to wait for 19 updates.

The Gutenberg plugin is uniquely positioned and can help us provide early feedback. No need to ship the React compiler in the WP until it's stable.

Should we open a discussion and ping the core teams?

Post:  https://react.dev/blog/2024/10/21/react-compiler-beta-release

**Comment by tyxla** [comment]
> The React Compiler team just released beta versions and is looking for feedback from early adopters. It also supports React 17+, so we don't have to wait for 19 updates.
> 
> The Gutenberg plugin is uniquely positioned and can help us provide early feedback. No need to ship the React compiler in the WP until it's stable.
> 
> Should we open a discussion and ping the core teams?
> 
> Post: https://react.dev/blog/2024/10/21/react-compiler-beta-release

Thanks for the heads up @Mamaduka 👍 I'll update to the beta soon and take it from there. 

I'm happy to start a discussion with the React core team, but in the meantime, I'd like to fix whatever we can before reaching out. A few examples come to mind:

- No longer using the `eslint-disable` directive whenever possible: #66324
- Using `react` directly instead of `@wordpress/element`: https://github.com/WordPress/gutenberg/discussions/66063
- Not passing hooks as a reference (we do that extensively with the command menu).

**Comment by tyxla** [comment]
Just updated to the beta version.

**Comment by Mamaduka** [comment]
@tyxla, 100% we should fix items you listed + any errors catched here by the linter.

**Comment by tyxla** [comment]
Indeed. I've already started fixing some of the violations and will be working on more:

- #66324
- #66327
- #66331

**Comment by tyxla** [comment]
Started experimenting with the React Compiler Babel plugin in https://github.com/WordPress/gutenberg/pull/66361

**Comment by Mamaduka** [comment]
@tyxla, you probably know this, but going to leave this tip for the future.

Running `npm run lint:js -- --quiet` helps filter out noise from ESLint warnings and only reports errors.

**Comment by tyxla** [comment]
> @tyxla, you probably know this, but going to leave this tip for the future.
> 
> Running `npm run lint:js -- --quiet` helps filter out noise from ESLint warnings and only reports errors.

Yep, I've been using the quiet flag a lot, there's so much noise from the disabled tests and the exhaustive deps violations 😕 

Thanks for the tip anyway @Mamaduka!

**Comment by tyxla** [comment]
Down from to **74** to **47** errors after the recent fixes 👍 

Thanks a lot for driving a bunch of those @Mamaduka 🙌

**Comment by tyxla** [comment]
⬇️ Down to **31** errors now.

**Comment by swissspidy** [comment]
I noticed that with React Compiler all `translators` comments are removed, which is not great, as WordPress translators rely on these comments for additional context. We must preserve them. I filed https://github.com/facebook/react/issues/31364 for this.

**Comment by tyxla** [comment]
> I noticed that with React Compiler all `translators` comments are removed, which is not great, as WordPress translators rely on these comments for additional context. We must preserve them. I filed [facebook/react#31364](https://github.com/facebook/react/issues/31364) for this.

Thanks @swissspidy. This issue belongs to the React Compiler itself, which you've rightfully also [reported there](https://github.com/WordPress/gutenberg/pull/66361#issuecomment-2439909143).

**Comment by tyxla** [comment]
⬇️ Down to **25** errors now.

**Comment by Mamaduka** [comment]
> ⬇️ Down to 25 errors now.

Most of those are related to Commands APIs hook style registration.

**Comment by tyxla** [comment]
> > ⬇️ Down to 25 errors now.
> 
> Most of those are related to Commands APIs hook style registration.

Since I'd like to avoid API changes, I've prepared a workaround for this in #66787. Gets the errors down from 24 to 8 🥳

**Comment by youknowriad** [comment]
When I introduced the hook API, I didn't see any other alternative. That said, it's clear (for multiple reasons) that it's not a perfect API. I would love if we could think of a better one in the future.

**Comment by tyxla** [comment]
> When I introduced the hook API, I didn't see any other alternative. That said, it's clear (for multiple reasons) that it's not a perfect API. I would love if we could think of a better one in the future.

I totally agree that a new API is the best future solution. For now, though, I'd prefer to keep the changes contained.

**Comment by tyxla** [comment]
#66787 landed, we're ⬇️ down to **8** errors now.

#66807 should fix 4 of those errors.

**Comment by tyxla** [comment]
⬇️ Only **2** errors left to go:

Screenshot 2024-11-10 at 6 34 47

**Comment by tyxla** [comment]
Thanks for working on those fixes @Mamaduka 🚀 🙇 

Unless something sneaks in, this should be the last fix before we enable the ESLint plugin: https://github.com/WordPress/gutenberg/pull/67085

**Comment by Mamaduka** [comment]
@tyxla, what do you think about merging the new link rules? Are there any possible downsides?

**Comment by tyxla** [comment]
@Mamaduka, this one is ready for review - please take a look when you have a chance. 

> @tyxla, what do you think about merging the new link rules? Are there any possible downsides?

Sorry @Mamaduka, which ones do you mean? In any case, I'd like to land this one first, and then we can tinker with any customizations in subsequent PRs.

In the meantime, in https://github.com/WordPress/gutenberg/pull/61788/commits/3bfdcd2f7e07260b7106e011bd722fb14089465d I've moved the ESLint config to the project level instead of the package since having a beta plugin in the package is debatable and could create some maintenance issues.

**Review by Mamaduka — APPROVED** [review]
> Sorry @Mamaduka, which ones do you mean? In any case, I'd like to land this one first, and then we can tinker with any customizations in subsequent PRs.

This is mostly a general question, in case you've spotted something on the React repo that warrants delaying a merge.

The ESLint plugin seems stable enough for merging ✅

**Review by gziolo — APPROVED** [review]
I don't have enough insights on wether version 19.0.0-beta-8a03594-20241020 is stable enough. The code changes itself look great. Excellent work on enabling all these rules inside Gutenberg repository.

> In the meantime, in https://github.com/WordPress/gutenberg/commit/3bfdcd2f7e07260b7106e011bd722fb14089465d I've moved the ESLint config to the project level instead of the package since having a beta plugin in the package is debatable and could create some maintenance issues.

Makes perfect sense. There is also this challenge that the ESLint plugin targets React 19. It doesn't seem to be a problem here, but you never know with wider set of projects.

**Comment by tyxla** [comment]
Thank you @Mamaduka and @gziolo 🙌 

> I don't have enough insights on wether version 19.0.0-beta-8a03594-20241020 is stable enough. The code changes itself look great. Excellent work on enabling all these rules inside Gutenberg repository.

There's a newer beta, and there's one every few days. I'm planning to keep the plugin up to date in subsequent PRs, but I want to enable it so folks stop introducing violations.

> There is also this challenge that the ESLint plugin targets React 19. It doesn't seem to be a problem here, but you never know with wider set of projects.

Yes, once React 19 actually gets released, this will be a bigger issue to consider.

**Comment by tyxla** [comment]
Bumping to the latest version here: https://github.com/WordPress/gutenberg/pull/67106

**Pull request merged**
